File: js/src/path-sum.js
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */

const tree = {
    val: 0,
    left: {
        val: 1,
        left: {
            val: 2,
            left: {
                val: 3,
            },
        },
        right: {
            val: 4,
        },
    },
    right: {
        val: 5,
    },
};

var hasPathSum = function (node, targetSum) {
    if (!node) {
        return false;
    }

    if (!node.left && !node.right && targetSum === node.val) {
        return true;
    }

    return hasPathSum(node.left, targetSum - node.val) || hasPathSum(node.right, targetSum - node.val);
};
console.log(hasPathSum(tree, 5));

File: rs/src/path_sum.rs

use std::rc::Rc;
use std::cell::RefCell;

pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

#[allow(dead_code)]
pub fn has_path_sum(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> bool {

    if let Some(node) = root {

        if node.borrow().left.is_none() && node.borrow().right.is_none() && target_sum == node.borrow().val {
            return true;
        }

        return has_path_sum(node.borrow().left.clone(), target_sum - node.borrow().val) || has_path_sum(node.borrow().right.clone(), target_sum - node.borrow().val)
    }


    return false;

}


#[test]
fn internal() {
    let node = TreeNode {
        val: 1,
        left: Some(Rc::new(RefCell::new(TreeNode {
            val: 2,
            left: None,
            right: None,
        }))),
        right: Some(Rc::new(RefCell::new(TreeNode {
            val: 3,
            left: None,
            right: None,
        }))),
    };
    let n = has_path_sum(Some(Rc::new(RefCell::new(node))), 4);
    assert_eq!(n, true);
}
